x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ClientHelper.java

```diff
@@ -77,6 +84,85 @@ public static void assertNoAuthorizationHeader(Map<String, String> headers) {
+        return maybeRewriteAuthenticationHeadersForVersion(
+            filterSecurityHeaders(threadContext.getHeaders()),
+            key -> new AuthenticationContextSerializer(key).readFromContext(threadContext),
+            minNodeVersion
+        );
+    }
+
+    /**
+     * Similar to {@link #getPersistableSafeSecurityHeadersForVersion(ThreadContext, Version)}, but works on a Map of headers instead of
+     * ThreadContext.
+     */
+    public static Map<String, String> getPersistableSafeSecurityHeadersForVersion(Map<String, String> headers, Version minNodeVersion) {
+        final CheckedFunction<String, Authentication, IOException> authenticationReader = key -> {
+            final String authHeader = headers.get(key);
+            return authHeader == null ? null : AuthenticationContextSerializer.decode(authHeader);
+        };
+        return maybeRewriteAuthenticationHeadersForVersion(filterSecurityHeaders(headers), authenticationReader, minNodeVersion);
+    }
+
+    private static Map<String, String> maybeRewriteAuthenticationHeadersForVersion(
+        Map<String, String> filteredHeaders,
+        CheckedFunction<String, Authentication, IOException> authenticationReader,
+        Version minNodeVersion
+    ) {
+        Map<String, String> newHeaders = null;
+
+        final String authHeader = getSingleRewrittenAuthenticationHeader(
+            authenticationReader,
+            AuthenticationField.AUTHENTICATION_KEY,
+            minNodeVersion
+        );
+        if (authHeader != null) {
+            newHeaders = new HashMap<>();
+            newHeaders.put(AuthenticationField.AUTHENTICATION_KEY, authHeader);
+        }
+
+        final String secondaryHeader = getSingleRewrittenAuthenticationHeader(
+            authenticationReader,
+            SecondaryAuthentication.THREAD_CTX_KEY,
+            minNodeVersion
+        );
+        if (secondaryHeader != null) {
+            if (newHeaders == null) {
+                newHeaders = new HashMap<>();
+            }
+            newHeaders.put(SecondaryAuthentication.THREAD_CTX_KEY, secondaryHeader);
+        }
+
+        if (newHeaders != null) {
+            final HashMap<String, String> mutableHeaders = new HashMap<>(filteredHeaders);
+            mutableHeaders.putAll(newHeaders);
+            return Map.copyOf(mutableHeaders);
+        } else {
+            return filteredHeaders;
+        }
+    }
+
+    private static String getSingleRewrittenAuthenticationHeader(
+        CheckedFunction<String, Authentication, IOException> authenticationReader,
+        String authenticationHeaderKey,
+        Version minNodeVersion
+    ) {
+        try {
+            final Authentication authentication = authenticationReader.apply(authenticationHeaderKey);
+            if (authentication != null && authentication.getVersion().after(minNodeVersion)) {
+                return authentication.maybeRewriteForOlderVersion(minNodeVersion).encode();
```

Comment:
This works for now. In a future PR, I intend to review how Authentication is constructed, serialised and downgraded.